docs/API_DESIGN.md

```diff
@@ -0,0 +1,443 @@
+# tbDEX API Design (APID) <!-- omit in toc -->
+
+**Last Updated:** June 13, 2024
+
+**Custom DSL Version:** 0.1.0
+
+- [Web5 Dependencies](#web5-dependencies)
+- [Resources](#resources)
+  - [`ResourceKind`](#resourcekind)
+  - [`ResourceMetadata`](#resourcemetadata)
+  - [`Offering`](#offering)
+    - [`OfferingData`](#offeringdata)
+    - [`PayinDetails`](#payindetails)
+    - [`PayinMethod`](#payinmethod)
+    - [`PayoutDetails`](#payoutdetails)
+    - [`PayoutMethod`](#payoutmethod)
+  - [`Balance`](#balance)
+    - [`BalanceData`](#balancedata)
+- [Messages](#messages)
+  - [`MessageKind`](#messagekind)
+  - [`MessageMetadata`](#messagemetadata)
+  - [`Rfq`](#rfq)
+    - [`CreateRfqData`](#createrfqdata)
+    - [`CreateSelectedPayinMethod`](#createselectedpayinmethod)
+    - [`CreateSelectedPayoutMethod`](#createselectedpayoutmethod)
+    - [`RfqData`](#rfqdata)
+    - [`SelectedPayinMethod`](#selectedpayinmethod)
+    - [`SelectedPayoutMethod`](#selectedpayoutmethod)
+    - [`RfqPrivateData`](#rfqprivatedata)
+    - [`PrivatePaymentDetails`](#privatepaymentdetails)
+  - [`Quote`](#quote)
+    - [`QuoteData`](#quotedata)
+    - [`QuoteDetails`](#quotedetails)
+    - [`PaymentInstructions`](#paymentinstructions)
+  - [`Order`](#order)
+  - [`OrderStatus`](#orderstatus)
+    - [`OrderStatusData`](#orderstatusdata)
+  - [`Close`](#close)
+    - [`CloseData`](#closedata)
+- [HTTP Client](#http-client)
+  - [`Exchange`](#exchange)
+  - [`get_offerings()`](#get_offerings)
+  - [`get_balances()`](#get_balances)
+  - [`create_exchange()`](#create_exchange)
+  - [`submit_order()`](#submit_order)
+  - [`submit_close()`](#submit_close)
+  - [`get_exchange()`](#get_exchange)
+  - [`get_exchanges()`](#get_exchanges)
+
+> [!WARNING]
+>
+> We need to define `JsonNode`
```

Comment:
I Will think about this... there must be a idiomatic way

Reply:
Right now I'm referencing the [Custom DSL over in `web5-rs`](https://github.com/TBD54566975/web5-rs/blob/main/docs/CUSTOM_DSL.md), even though we're inside `tbdex-rs`. Basically I think we just need to add the concept of a _JSON Object_ type to the [Custom DSL](https://github.com/TBD54566975/web5-rs/blob/main/docs/CUSTOM_DSL.md) and then update this APID and [the `web5-rs` APID](https://github.com/TBD54566975/web5-rs/blob/main/docs/API_DESIGN.md) to use that new syntax.